### PR TITLE
[Enhance] make re2 driver-local to avoid contention

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

`re2::RE2` is shared among all threads/drivers, which could introduce contention on the `re2::DFA::cache_mutex`.

The solution is to make it `pipeline-driver local`, to avoid any contention.

According to benchmark on a many-cores machine:
- SQL: `SELECT REGEXP_REPLACE(Referer, '^https?://(?:www\.)?([^/]+)/.*$', '\1') AS k, AVG(length(Referer)) AS l, COUNT(*) AS c, MIN(Referer) FROM hits WHERE Referer <> '' GROUP BY k HAVING COUNT(*) > 100000 ORDER BY l DESC LIMIT 25;`
- Before: 8.04s
- After: 1.07s

### CODE REVIEW

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.3

File: be/src/runtime/current_thread.h [hunk 1]
```diff
@@ -55,6 +55,7 @@ class CurrentThread {
         _fragment_instance_id = fragment_instance_id;
     }
     void set_pipeline_driver_id(int32_t driver_id) { _driver_id = driver_id; }
+    int32_t get_driver_id() const { return _driver_id; }
```
Comment (stdpain) [inline be/src/runtime/current_thread.h:58]: Did we assign driver id in pip_scan_io thread ?

File: be/src/exprs/vectorized/string_functions.cpp [hunk 2]
```diff
@@ -2517,16 +2519,35 @@ int StringFunctions::index_of(const char* source, int source_count, const char*
...
     StringFunctionsState() : regex(), options() {}
+
+    // Implement a driver-local regex, to avoid lock contention on the RE2::cache_mutex
+    re2::RE2* get_or_prepare_regex() {
+        DCHECK(const_pattern);
+        int32_t driver_id = CurrentThread::current().get_driver_id();
+        if (driver_id == 0) {
+            return regex.get();
+        }
+        auto iter = driver_regex_map.find(driver_id);
```
Comment (stdpain) [inline be/src/exprs/vectorized/string_functions.cpp:2538]: Is it safe call find without a mutex ?

Comment (mofeiatwork) [reply be/src/exprs/vectorized/string_functions.cpp:2538]: Mistake. I should use the thread-safe version.

File: be/src/runtime/current_thread.h (hunk 1 above)
Comment (mofeiatwork) [reply be/src/runtime/current_thread.h:58]: Fixed. Share the same driver_id in io-thread.

File: be/src/exprs/vectorized/string_functions.cpp [hunk 3]
```diff
@@ -2517,16 +2520,37 @@ int StringFunctions::index_of(const char* source, int source_count, const char*
...
+    std::string pattern;
     std::unique_ptr<re2::RE2> regex;
     std::unique_ptr<re2::RE2::Options> options;
     bool const_pattern{false};
+    DriverMap driver_regex_map; // regex for each pipeline_driver, to make it driver-local
 
     StringFunctionsState() : regex(), options() {}
+
+    // Implement a driver-local regex, to avoid lock contention on the RE2::cache_mutex
+    re2::RE2* get_or_prepare_regex() {
```
Comment (kangkaisen) [inline be/src/exprs/vectorized/string_functions.cpp:2536]: If non-pipeline engine run here, is the result right?

Comment (mofeiatwork) [reply be/src/exprs/vectorized/string_functions.cpp:2536]: It would use the default regex, which is same as before.

**Review by kangkaisen — APPROVED** [review]

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by kangkaisen** [comment]
run starrocks_admit_test

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**